ext/afform/core/CRM/Afform/ArrayHtml.php

```diff
@@ -371,7 +371,7 @@ protected function decodeAttrValue($type, $txtAttrValue) {
    * @return string
    */
   public function replaceUnicodeChars($markup) {
-    return mb_convert_encoding($markup, 'HTML-ENTITIES', 'UTF-8');
+    return htmlspecialchars_decode(htmlentities($markup, ENT_COMPAT, 'utf-8', false));
```

Comment:
This only appears to do half of what the original intended. It leaves things like `&` alone, but doesn't convert unicode chars to htmlentities. For example:

`php -r "echo mb_convert_encoding('a b c д e', 'HTML-ENTITIES', 'UTF-8');"` gives `a b c &#1076; e`

but `php -r "echo htmlentities('a b c д e', ENT_COMPAT, 'utf-8', false);"` gives `a b c д e`, and applying htmlspecialchars_decode thus also gives `a b c д e`.

I'm not sure what the right replacement is, assuming this is actually needed. I'm not clear how it's used.

Reply:
@demeritcowboy this seem to be the original commit https://github.com/civicrm/civicrm-core/commit/3a9e5a624237332d69b063f2bc7c816e7ebf9394 and has a reference to this stack exchange https://stackoverflow.com/questions/59215211/why-does-domdocument-mess-up-unicode-nonbreaking-spaces @colemanw do you want to comment on this?

Reply:
I'm still not clear exactly how it gets used, but based on that and the other history it seems like the goal is to make it clear on screen when a character is not as it seems, both for readability and to avoid overwriting it. For example `nbsp` looks like a regular space, so is easy to mistake/overwrite, so `&nbsp;` is better.

So if that's true, then this replacement is good for nbsp and some others. But for example there are similar unicode chars where it doesn't do the same replacement. An example you can clearly see (ahem) here in [this line](https://github.com/civicrm/civicrm-core/blob/f7a1f0d9e621bd14976abb512595fb3ca43acd09/tests/phpunit/Civi/Core/FormatTest.php#L193). Clearly that is not a space or nbsp but is a nnbsp. Previously the mb_convert would output `&#8239;`. Now it outputs as the original unicode char.

So is that a blocker? I'm not sure. If this function is just about readability/accident-prevention and not an absolute requirement, then it's not a blocker and is just a nice-to-have. If it's truly required, then one alternative might be to have it output the numeric entity, e.g. `&#8239;`, for any char in the string that is not ascii or in a fixed list of chars.